### DISCUSSION TITLE
message consumed in k8s env

### DISCUSSION THREAD

**Discussion by janiu-001** [rabbitmq/rabbitmq-server#5804] [Other]
Hi Expert,

Very apreciated for your answer on my question.

I deloyed the rabbitmq cluster(2 nodes)  and a consume service(3 pods) in k8s env on my own pc.

In my testing, the message could be sent to the two rabbitmq with round roubin, but all the 3 pod of the consumer received the messages.

By the way, i used the mqtt protocol by enabling rabbitmq_mqtt protocol. How to ensure only one pod received the message and keep the consume order, Otherwise, it will lead to repeated consumption.  Do we have the concept of group like kafka.  Look forward for your response. Follwong is my code, connect the mqtt with the k8s service name :

![image](https://user-images.githubusercontent.com/19970581/190880003-57a31c96-082d-4b4d-bf42-2f28e502efa4.png)

**Comment by michaelklishin** [comment]
Consider using GitHub Discussions and not issues for questions.

You haven't provided any details as to how RabbitMQ was deployed, or what version is used, or why you decided to [use an configuration the docs explicitly recommend against](https://www.rabbitmq.com/clustering.html#node-count). But I guess in this case these are less important in this case.

With MQTT 3.1, which RabbitMQ implements, every client gets its own queue. So there won't be a round-robin dispatch demonstrated in [tutorial 2](https://www.rabbitmq.com/getstarted.html). MQTT also does not have the concept of a [single active consumer](https://www.rabbitmq.com/consumers.html#single-active-consumer) like AMQP 0-9-1, AMQP 1.0 (to an extent) and RabbitMQ Stream protocol.

MQTT 5 has a concept of SAC called "shared subscriptions". Its support will be coming to RabbitMQ in the next 6 or so months. In the meantime, you can use one of the protocols mentioned above.

And please do not use two-node clusters.

**Reply by janiu-001** [reply]
i deployed the rabbitmq with two-node with helm(bitname). Yeach, i remembered that Rabbitmq does not support two nodes. It only supports the odd cluter

Yes i know that  "shared subscriptions", it is like the group in Kafka. But only supproted by MQTT5, since most of our device is using MQTT3, so is there any way to meet my needs for MQTT 3.x? or Should i only have one pods?  My  goal is to deploy my consume service in k8s env, it may have several pods for load balance, but in order to  avoid the repeated consumation, it is better to ensure the messages only be consumed by one pod

**Reply by michaelklishin** [reply]
There is no concept of shared consumers (which is called "competing consumers" in messaging and is demonstrated in RabbitMQ [tutorial 2](https://www.rabbitmq.com/getstarted.html)).

So you have to wait until MQTTv5 support ships or use a different protocol like in the tutorial above.
